broker/src/main/java/org/apache/rocketmq/broker/util/HookUtils.java

```diff
@@ -122,10 +124,85 @@ public static PutMessageResult handleScheduleMessage(BrokerController brokerCont
         final MessageExtBrokerInner msg) {
         final int tranType = MessageSysFlag.getTransactionValue(msg.getSysFlag());
         if (tranType == MessageSysFlag.TRANSACTION_NOT_TYPE
-            || tranType == MessageSysFlag.TRANSACTION_COMMIT_TYPE) {
-            if (!TopicValidator.RMQ_SYS_SCHEDULE_TOPIC.equals(msg.getTopic()) && msg.getDelayTimeLevel() > 0) {
-                transformDelayLevelMessage(brokerController, msg);
+                || tranType == MessageSysFlag.TRANSACTION_COMMIT_TYPE) {
+            if (!isRolledTimerMessage(msg)) {
+                if (checkIfTimerMessage(msg)) {
+                    if (!MessageStoreConfig.isTimerWheelEnable()) {
+                        //wheel timer is not enabled, reject the message
+                        return new PutMessageResult(PutMessageStatus.WHEEL_TIMER_NOT_ENABLE, null);
+                    }
+                    PutMessageResult tranformRes = transformTimerMessage(brokerController, msg);
+                    if (null != tranformRes) {
+                        return tranformRes;
+                    }
+                }
             }
+            // Delay Delivery
+            if (msg.getDelayTimeLevel() > 0) {
+                transformDelayLevelMessage(brokerController,msg);
+            }
+        }
+        return null;
+    }
+    private static boolean isRolledTimerMessage(MessageExtBrokerInner msg) {
+        return TimerMessageStore.TIMER_TOPIC.equals(msg.getTopic());
+    }
+
+    public static boolean checkIfTimerMessage(MessageExtBrokerInner msg) {
+        if (msg.getDelayTimeLevel() > 0) {
+            if (null != msg.getProperty(MessageConst.PROPERTY_TIMER_DELIVER_MS)) {
+                MessageAccessor.clearProperty(msg, MessageConst.PROPERTY_TIMER_DELIVER_MS);
+            }
+            if (null != msg.getProperty(MessageConst.PROPERTY_TIMER_DELAY_SEC)) {
+                MessageAccessor.clearProperty(msg, MessageConst.PROPERTY_TIMER_DELAY_SEC);
+            }
+            return false;
+            //return this.defaultMessageStore.getMessageStoreConfig().isTimerInterceptDelayLevel();
+        }
+        //double check
+        if (TimerMessageStore.TIMER_TOPIC.equals(msg.getTopic()) || null != msg.getProperty(MessageConst.PROPERTY_TIMER_OUT_MS)) {
+            return false;
+        }
+        return null != msg.getProperty(MessageConst.PROPERTY_TIMER_DELIVER_MS) || null != msg.getProperty(MessageConst.PROPERTY_TIMER_DELAY_MS) || null != msg.getProperty(MessageConst.PROPERTY_TIMER_DELAY_SEC);
+    }
+    private static PutMessageResult transformTimerMessage(BrokerController brokerController, MessageExtBrokerInner msg) {
+        //do transform
+        int delayLevel = msg.getDelayTimeLevel();
+        long deliverMs;
+        try {
+            if (msg.getProperty(MessageConst.PROPERTY_TIMER_DELAY_SEC) != null) {
+                deliverMs = System.currentTimeMillis() + Integer.parseInt(msg.getProperty(MessageConst.PROPERTY_TIMER_DELAY_SEC)) * 1000;
```

Comment:
when set setDelayTimeSec(24 * 3600 * 3), integer overflows

Reply:
thx, it has been resolved.